aiodataloader.py

```diff
@@ -1,29 +1,82 @@
-from asyncio import gather, ensure_future, get_event_loop, iscoroutine, iscoroutinefunction
-from collections import namedtuple
-from collections.abc import Iterable
+from asyncio import (
+    AbstractEventLoop,
+    Future,
+    gather,
+    ensure_future,
+    get_event_loop,
+    iscoroutine,
+    iscoroutinefunction,
+)
 from functools import partial
 
-from typing import List  # flake8: noqa
+from typing import (
+    Any,
+    Callable,
+    Coroutine,
+    Dict,
+    Generic,
+    Iterable,
+    Iterator,
+    List,
+    Optional,
+    TypeVar,
+    Union,
+)
+from typing_extensions import Protocol, TypeGuard
+
+__version__ = "0.2.1"
+
+
+def iscoroutinefunctionorpartial(fn: Callable) -> TypeGuard[Callable[..., Coroutine]]:
+    return iscoroutinefunction(fn.func if isinstance(fn, partial) else fn)
 
 
-__version__ = '0.2.1'
+KeyT = TypeVar("KeyT")
+ReturnT = TypeVar("ReturnT")
+CacheKeyT = TypeVar("CacheKeyT")
+T = TypeVar("T")
 
-Loader = namedtuple('Loader', 'key,future')
 
+class BatchLoadFnProto(Protocol[KeyT, ReturnT]):
+    async def __call__(self, keys: List[KeyT]) -> List[ReturnT]:
+        ...
 
-def iscoroutinefunctionorpartial(fn):
-    return iscoroutinefunction(fn.func if isinstance(fn, partial) else fn)
 
+class Loader(tuple, Generic[KeyT, ReturnT]):
```

Comment:
Why is it actually necessary for Loader to be generic?  I think the original `namedtuple('Loader', 'key,future')` can still be used, since this is only used internally.

I think its a good goal for this PR to have zero runtime impact, to ensure no possibility of runtime breakage.

Reply:
Not keeping it Generic loses typing information through _queue, the generic allows it to be inferred throughout the usage in DataLoader. It should mostly only matter inside DataLoader itself, but I haven't prodded it hard enough to be 100% sure about that.

Reply:
Not having it Generic does lose information, but as far as I can tell that type information isn't needed to satisfy the goal of this change, which is to make `DataLoader` fully typed.

Event though this change seems benign, I think the goal of this change should be to have absolutely no runtime changes whatsoever in the code, to avoid any possibility of breakage in the wild.  There could be integrations that depend somehow on Loader being a NamedTuple.

If we change the runtime behavior of the library, I think we will have to do a major release, and ideally we can do this as a point release.

Reply:
Yeah, I'm fine with it, it's always easy to tweak later if someone actually runs into it

Reply:
Worth noting that while taking it out I was reminded that 

```
dispatch_queue_batch
failed_dispatch
```

both use the generic in their signature, and these have public function naming convention even though I can't fathom a reason why they'd actually be called publicly.